Sources/SPMBuildCore/PluginInvocation.swift

```diff
@@ -64,45 +61,220 @@ extension PluginTarget {
+        let initialMessage: Data
         do {
-            inputStruct = try serializer.makePluginScriptRunnerInput(
-                rootPackage: package,
-                pluginWorkDir: outputDirectory,
-                toolSearchDirs: toolSearchDirectories,
-                toolNamesToPaths: toolNamesToPaths,
-                pluginAction: action)
+            var serializer = PluginContextSerializer(buildEnvironment: buildEnvironment)
+            let pluginWorkDirId = try serializer.serialize(path: outputDirectory)
+            let toolSearchDirIds = try toolSearchDirectories.map{ try serializer.serialize(path: $0) }
+            let toolNamesToPathIds = try toolNamesToPaths.mapValues{ try serializer.serialize(path: $0) }
+            let actionMessage: HostToPluginMessage
+            switch action {
+                
+            case .createBuildToolCommands(let package, let target):
+                let rootPackageId = try serializer.serialize(package: package)
+                guard let targetId = try serializer.serialize(target: target) else {
+                    throw StringError("unexpectedly was unable to serialize target \(target)")
+                }
+                let wireInput = WireInput(
+                    paths: serializer.paths,
+                    targets: serializer.targets,
+                    products: serializer.products,
+                    packages: serializer.packages,
+                    pluginWorkDirId: pluginWorkDirId,
+                    toolSearchDirIds: toolSearchDirIds,
+                    toolNamesToPathIds: toolNamesToPathIds)
+                actionMessage = .createBuildToolCommands(
+                    context: wireInput,
+                    rootPackageId: rootPackageId,
+                    targetId: targetId)
+            
+            case .performCommand(let package, let arguments):
+                let rootPackageId = try serializer.serialize(package: package)
+                let wireInput = WireInput(
+                    paths: serializer.paths,
+                    targets: serializer.targets,
+                    products: serializer.products,
+                    packages: serializer.packages,
+                    pluginWorkDirId: pluginWorkDirId,
+                    toolSearchDirIds: toolSearchDirIds,
+                    toolNamesToPathIds: toolNamesToPathIds)
+                actionMessage = .performCommand(
+                    context: wireInput,
+                    rootPackageId: rootPackageId,
+                    arguments: arguments)
+            }
+            initialMessage = try actionMessage.toData()
         }
         catch {
             return callbackQueue.async { completion(.failure(PluginEvaluationError.couldNotSerializePluginInput(underlyingError: error))) }
         }
+        
+        // Handle messages and output from the plugin.
+        class ScriptRunnerDelegate: PluginScriptRunnerDelegate {
```

Comment:
This code is hoisted up from the DefaultPluginScriptRunner, which now deals only with the mechanics of invoking the subprocess and sending messages, and not with the format of the messages.  So there is now a PluginScriptRunnerDelegate which handles callbacks about output and messages received, and that in turn calls back to the PluginInvocationDelegate, which is what the caller sees and has callbacks specific to plugin actions.

This way the PluginScriptRunner is restored to what it originally was, which is just a way of optionally compiling and then calling a subprocess and sending messages to it on a particular platform.